File: backend/app/services/member_service.py
import re
import time

from fastapi import BackgroundTasks, HTTPException, status

from app.db.client import supabase
from app.middleware.auth import CurrentUser
from app.models.member import MemberCreateIn, MemberOut, SimpleLoginIn, SimpleLoginOut

_BIRTHDAY_RE = re.compile(r"^(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$")

# In-memory throttle on simple-login attempts per email — there's no password
# left to guess, but without this, the endpoint would let anyone script
# through a list of emails to discover which ones belong to real members.
# A single Render instance is fine for this; resets on restart/deploy, which
# is an acceptable tradeoff for a login-attempt limiter, not a security
# control that needs to survive that.
_LOGIN_ATTEMPTS: dict[str, list[float]] = {}
_MAX_ATTEMPTS = 5
_WINDOW_SECONDS = 600


def _check_rate_limit(email: str) -> None:
    now = time.monotonic()
    attempts = [t for t in _LOGIN_ATTEMPTS.get(email, []) if now - t < _WINDOW_SECONDS]
    if len(attempts) >= _MAX_ATTEMPTS:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail="Too many attempts. Please wait a few minutes and try again.",
        )
    attempts.append(now)
    _LOGIN_ATTEMPTS[email] = attempts


async def register(body: MemberCreateIn, background_tasks: BackgroundTasks) -> MemberOut:
    """Public self-registration from the login screen. Reuses the same
    create-member logic admins use, scoped to the club's own club_id since
    Phase 1 only supports one club."""
    from app.db import admin_members as admin_db
    from app.services import admin_member_service

    club_id = await admin_db.get_default_club_id()
    if not club_id:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Club not configured")
    return await admin_member_service.create_member(club_id, body, background_tasks)


async def simple_login(body: SimpleLoginIn) -> SimpleLoginOut:
    """Passwordless sign-in: email is the only thing that gates entry — name
    and phone are recorded/kept fresh but not required to match. Deliberately
    excludes super_admin accounts, which keep signing in with a password."""
    from app.db import members as db

    email = body.email.strip().lower()
    _check_rate_limit(email)

    member = await db.get_by_email(email)
    if not member or member.get("is_guest") or not member.get("auth_user_id"):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No account found with this email. Please register first.",
        )
    if not member.get("is_active", True):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="This account has been deactivated.")
    if member.get("app_role") == "super_admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Super admin accounts sign in with a password.",
        )

    await db.update_contact_info(member["id"], body.name.strip(), body.phone.strip())

    link = supabase.auth.admin.generate_link({"type": "magiclink", "email": email})
    return SimpleLoginOut(email=email, hashed_token=link.properties.hashed_token)


async def get_me(user: CurrentUser) -> MemberOut:
    from app.db import members as db
    row = await db.get_by_auth_user_id(user.id)
    if not row:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Member record not found")
    return MemberOut(**row)


async def get_club_members(user: CurrentUser) -> list:
    from app.db import members as db
    rows = await db.get_club_members(user.club_id)
    return rows


async def get_my_stats(user: CurrentUser) -> dict:
    from app.db import members as db
    return await db.get_stats_by_auth_user_id(user.id)


async def confirm_password_changed(user: CurrentUser) -> None:
    from app.db import admin_members as admin_db
    await admin_db.clear_must_change_password(user.id)


async def update_birthday(birthday: str, _user: CurrentUser) -> MemberOut:
    if not _BIRTHDAY_RE.match(birthday):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Birthday must be in MM-DD format",
        )
    # Atomic UPDATE: set birthday = birthday, birthday_collected = true WHERE id = member.id
    # TODO: implement
    raise NotImplementedError

File: web/src/services/memberService.ts
import { apiRequest } from '@/lib/apiClient';
import type { Member, ClubRole, AppRole, MemberInitials } from '@/types';

export interface MemberCreatePayload {
  name: string;
  email: string;
  phone: string;
  birthday?: string; // MM-DD
  initials: MemberInitials;
}

export interface MemberUpdatePayload {
  name: string;
  email: string;
  phone: string;
  birthday?: string | null; // MM-DD
  initials: MemberInitials;
}

// ── Admin member management ────────────────────────────────────────────────

export function getAllMembers(token: string): Promise<Member[]> {
  return apiRequest<Member[]>('/admin/members', { token });
}

export function getMemberById(memberId: string, token: string): Promise<Member> {
  return apiRequest<Member>(`/admin/members/${memberId}`, { token });
}

export function createMember(payload: MemberCreatePayload, token: string): Promise<Member> {
  return apiRequest<Member>('/admin/members', { method: 'POST', body: payload, token });
}

// Public — no auth. Self-registration from the login screen.
export function registerMember(payload: MemberCreatePayload): Promise<Member> {
  return apiRequest<Member>('/members/register', { method: 'POST', body: payload });
}

export interface SimpleLoginPayload {
  email: string;
  name: string;
  phone: string;
}

export interface SimpleLoginResult {
  email: string;
  hashed_token: string;
}

// Public — no auth. Passwordless sign-in: email gates entry, name/phone are
// recorded but don't have to match exactly. Returns a token the caller
// exchanges via supabase.auth.verifyOtp({ email, token_hash, type: 'magiclink' }).
export function simpleLogin(payload: SimpleLoginPayload): Promise<SimpleLoginResult> {
  return apiRequest<SimpleLoginResult>('/members/simple-login', { method: 'POST', body: payload });
}

export function updateMemberDetails(memberId: string, payload: MemberUpdatePayload, token: string): Promise<Member> {
  return apiRequest<Member>(`/admin/members/${memberId}`, { method: 'PUT', body: payload, token });
}

export function resendInvite(memberId: string, token: string): Promise<void> {
  return apiRequest<void>(`/admin/members/${memberId}/resend-invite`, { method: 'POST', token });
}

export function setMemberActive(memberId: string, isActive: boolean, token: string): Promise<Member> {
  return apiRequest<Member>(`/admin/members/${memberId}/active`, {
    method: 'PUT',
    body: { is_active: isActive },
    token,
  });
}

export function updateMemberClubRole(memberId: string, clubRole: ClubRole, token: string): Promise<Member> {
  return apiRequest<Member>(`/admin/members/${memberId}/club-role`, {
    method: 'PUT',
    body: { club_role: clubRole },
    token,
  });
}

export function updateMemberAppRole(memberId: string, appRole: AppRole, token: string): Promise<Member> {
  return apiRequest<Member>(`/admin/members/${memberId}/app-role`, {
    method: 'PUT',
    body: { app_role: appRole },
    token,
  });
}

// ── Self / member ──────────────────────────────────────────────────────────

export function getMe(token: string): Promise<Member> {
  return apiRequest<Member>('/members/me', { token });
}

export function confirmPasswordChanged(token: string): Promise<void> {
  return apiRequest<void>('/members/me/confirm-password', { method: 'POST', token });
}

export function getMyStats(token: string): Promise<{ speeches: number; feedbacks: number }> {
  return apiRequest('/members/me/stats', { token });
}

export function getClubMembers(
  token: string,
): Promise<{ id: string; name: string; initials: MemberInitials; club_role: string; app_role: string | null; is_active: boolean }[]> {
  return apiRequest('/members/club', { token });
}

export function updateBirthday(
  birthday: string,
  token: string,
): Promise<Pick<Member, 'id' | 'birthday' | 'birthday_collected'>> {
  return apiRequest('/members/birthday', {
    method: 'PUT',
    body: { birthday },
    token,
  });
}
